### ISSUE TITLE
Request: Implement a Torznab Compatible API

### ISSUE THREAD

**Issue by CodeMonkey626** [nyaadevs/nyaa#18]
I would be great if nyaa.si could add a [Torznab](https://github.com/Sonarr/Sonarr/wiki/Implementing-a-Torznab-indexer) compatible API to query for torrents.

This would allow for direct integration with software such as [Sonarr](https://github.com/Sonarr/Sonarr) and [Radarr](https://github.com/Radarr/Radarr) as well ensuring applications don't need to resort to scraping to search for indexed torrents.

Background on Sonarr:

> Sonarr is a PVR for Usenet and BitTorrent users. It can monitor multiple RSS feeds for new episodes of your favorite shows and will grab, sort and rename them. It can also be configured to automatically upgrade the quality of files already downloaded when a better quality format becomes available.

Background on Torznab:

> The Newznab api specification is built around a simple rss feed with filtering and paging capabilities. It was designed for usenet primarily, but it can be easily adapted for torrents.

Torznab Documentation:
https://github.com/Sonarr/Sonarr/wiki/Implementing-a-Torznab-indexer

**Comment by Taloth** [comment]
:) I was posting an issue as well: copy/paste and I'll close the dupe:

----

I'm one of the Sonarr developers and I've been keeping an eye on the fallout of nyaa closing down.
With it, a lot of our users have been asking for alternatives. With some working rss feeds, but lacking search capabilities.

For that I'd like to propose that nyaa.si supports a torznab api. In essence it's an enhanced rss feed with additional attributes and paging support as well as query parameters to specify search parameters.
The Torznab api specification is based on Newznab. The Newznab spec is used by +90% of all usenet indexers and has provided a standardized way for several applications to communicate with all those usenet indexers, saving the apps the trouble of implementing an api client for each site separately.
Torznab is intended to serve a similar purpose for torrent indexers.

With nyaa.si being open-source, I'm sure I could find some contributors willing to invest some time. But that depends on whether the nyaa devs want it.

Please let me know if you're interested.

PS: The most basic search capability is based on free text search. Which should be easy enough to implement for a contributor, but has certain performance implications. Eventually indexing by (for example anidb id) & absolute episode numbers could seriously improve the performance. So it's a matter of "how far do you want to go".

**Comment by nyaadev** [comment]
I don't see a problem with adding this as far as it is possible.
Right now the rss search works exactly the same as the normal homepage search.
Adding the info we already have on the torrents to the rss is easy, but that's not a lot right now...

nyaatorrents never had metadata like anidb or even which anime / episode number.
It's was always fulltext search only.
In the future we want a system that is more like that of private trackers which proper metadata and search, but if, how or when that will be implemented is uncertain.

**Comment by xelra** [comment]
I'm very much looking forward to this. I guess all that's needed to get started is a text search with categories and serving it up in a torznab compatible rss response. So it'd pretty much be just like the search on the website.

Anything else, metadata and more details, can always be added later on when those features become available on the site.

I'm very glad that torznab might finally get some momentum in the torrent world.

**Comment by Taloth** [comment]
@xelra one of the things with torznab is that we need to map nyaa categories on the newznab cats. 5070 for Anime. And 10xxxx for site-specific categories. Using 10xxxx isn't strictly required, but it was the original 'custom' range specified by newznab and I attempted to stick as much to nn as possible.
For hd4free we mapped the site rss cats to both 10xxxx and 50xx categories (one release can have multiple categories).

Why I'm saying this is because we need to define a mapping that makes sense. Which of the nyaa categories should be exposed by 5070 and thus be the 'default' for users?

This is what I typed up earlier today: https://gist.github.com/Taloth/17c16edcbb438cb024d9dc68d0121434

Look at it and tell me what you think.

My plan is to work on the templates, explain how it should work and then hopefully get a volunteer python expert to do the coding.
I just want to get the show on the road, coz I don't have a lot of time considering the stuff I have to do for Sonarr.

**Comment by xelra** [comment]
I think the default category should be the same as with the old nyaa: Anime - English-translated, no remakes.

**Comment by Taloth** [comment]
Yes for the TV category. But Nyaa.si seems to have a lot other categories, and if we want it to adopt torznab it should work for all releases that it provides, not just to satisfy Sonarr.
One thing I'd like to know is whether the Nyaa.si categories are just draft, or will be tweaked. Take for example Literature, I guessed that it would mostly be Ebooks, but is it? Or does it involve comics, coz those have a separate category in nn. That's why I pinged you, hoped you'd be willing to sort it out. :smile:
And afaik Nyaa.si doesn't have a Remake category.

**Comment by xelra** [comment]
Hmm, I'm not knowledgable about the other categories since I never used them.

The remakes don't seem to be a category, but rather some kind of tag. Check out the web search. It's a filter.

**Comment by OmgImAlexis** [comment]
I'm one of the devs from [Medusa](https://github.com/pymedusa/Medusa/) and [Sickrage](https://github.com/sickrage/sickrage), we'll be keeping an eye on this.

**Comment by CodeMonkey626** [comment]
@OmgImAlexis would you or any other dev on the Medusa/Sickrage team be interesting in working on a PR for this feature? I believe the nyaadevs team is currently pretty busy with other items, so this may be an area where the community can step in and get this implemented sooner.

@Taloth is your caps draft something an initial implementation could be built on top of at this point?

**Comment by OmgImAlexis** [comment]
@p0psicles @labrys @medariox @sharkykh anyone up for this?

**Comment by Taloth** [comment]
@CodeMonkey626 Yes and no, that draft is distant history, look at the stuff I have so far https://github.com/Taloth/nyaa/tree/torznab_api. It was working reasonably with the mariadb db, but I'm updating it for elastic support so some fields are currently left it.
I need to talk to the nyaadevs on IRC. coz the different models is making the UI render layer unnecessarily complicated, imho. It would be nice if a wrapper is available around the search module that exposes an consistent model.

Please note that last time I worked with python was like a decade ago. So at one point it needs to be picked up by someone more experienced. I just wanted to create a PoC, showing the api calls that need to be handled, the caps and feed template.

Btw: That branch also contains a docker-compose file that will spin up nyaa, elastic and mariadb. But that's just for personal testing since I'm developing on windows. And needs a couple of docker exec commands to init the db and such.

**Comment by CodeMonkey626** [comment]
@Taloth Nice! You can find the nyaa devs on IRC at #nyaa-dev@irc.rizon.net. You'll want to talk to ReimuHakurei, Botox, AMM, and snow.

**Comment by CodeMonkey626** [comment]
@Taloth Just noticed there's some additional changes to the RSS template in flight: https://github.com/nyaadevs/nyaa/pull/51

**Comment by sharkykh** [comment]
@CodeMonkey626 
Yep, mine... Not sure why it didn't get picked up yet.

@OmgImAlexis 
Not sure if I'm up for it.
My dev environment is not really set up like the site (I'm using SQLite and no elastic search)

**Comment by Taloth** [comment]
Problem with that is that those additional elements aren't defined in the rss spec. That's why I put the size in the enclosure, but as element it's probably ok too. I know Sonarr looks for it (I'd recommend it be in bytes, not formatted).
However the seeders, leechers & downloads feel totally arbitrary. I doubt many clients understand that out of the box. ezrss defined the 'seeds' & 'peers' element inside a torrent element with xmlns, kat used the same spec but without the enclosing torrent element.
If you want the rss to include that information (and be usable by clients that historically could handle ezrss), best choice would be to tack that namespace on.

PS: most rss feeds had it as formatted text in the description.

@sharkykh If you have docker available and some spare memory, you can spin up the whole stack in there. The docker-compose file will do that, except for the db init (you can kick off those scripts manually) and binlog sync.

**Comment by labrys** [comment]
I don't have time to dev this, however I am willing to review and provide feedback and python guidance for anyone who does want to work on an implementation.  If you want me to look at what you've got, just tag me in your branch.
